ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/TransactionComponents.kt

```diff
@@ -0,0 +1,196 @@
+ */
+@CordaSerializable
+@DoNotImplement
+interface Attachment {
+
+    val id: SecureHash
+    val size: Int
+    val signatories: Set<PublicKey>
+
+    /**
+     * Finds the specified file (case insensitively) within the attachment and copies it to the specified output stream.
+     *
+     * @param path The path of the file to find within the attachment.
+     * @param output The output stream where the file should be copied to.
+     * @throws [FileNotFoundException] if the specified path does not exist in the attachment.
+     */
+    fun extractFile(path: String, output: OutputStream)
+
+    /**
+     * Opens the current [Attachment].
+     *
+     * @return Returns the [InputStream] for the current [Attachment].
+     */
+    fun open(): InputStream
+
+    /**
+     * Opens the current [Attachment] as a JAR (Java Archive).
+     *
+     * @return Returns the [JarInputStream] for the current [Attachment].
+     */
+    fun openAsJar(): JarInputStream {
+        val stream = open()
+        return try {
+            JarInputStream(stream)
+        } catch (ex: IOException) {
+            stream.use { throw ex }
+        }
+    }
+}
+
+/**
+ * Defines an interval on a timeline; not a single, instantaneous point.
+ *
+ * There is no such thing as "exact" time in distributed systems, due to the underlying physics involved, and other
+ * issues such as network latency. A time window represents an approximation of an instant with a margin of tolerance,
+ * and may be fully bounded, or unbounded towards negative or positive infinity.
+ *
+ * @property from The boundary at which the time window begins, or null if the time window is unbounded towards negative infinity.
+ * @property until The boundary at which the time window ends, or null if the time window is unbounded towards positive infinity.
+ * @property midpoint The midpoint between a fully bounded time window, or null if the time window is unbounded towards positive or negative infinity.
+ * @property duration The duration of a fully bounded time window, or null if the time window is unbounded towards positive or negative infinity.
+ */
+@DoNotImplement
+@CordaSerializable
+interface TimeWindow {
+    val from: Instant?
+    val until: Instant?
+    val midpoint: Instant?
+    val duration: Duration? get() = if (from == null || until == null) null else Duration.between(from, until)
+
+    /**
+     * Determines whether the current [TimeWindow] contains the specified [Instant].
+     *
+     * @param instant The [Instant] to check is contained within the current [TimeWindow].
+     * @return Returns true if the current [TimeWindow] contains the specified [Instant]; otherwise, false.
+     */
+    operator fun contains(instant: Instant): Boolean
+}
+
+/**
+ * Defines information that applies to a transaction state.
+ *
+ * @property contractId The class name of the [Contract] associated with the transaction state.
```

Comment:
class names are ambiguous in OSGi so this is not accurate. It will be an opaque tag that refers to the CPK main bundle plus class name.